File: src/main/java/cn/labzen/web/ap/internal/element/ElementParameter.java
package cn.labzen.web.ap.internal.element;

import cn.labzen.web.ap.internal.Utils;
import com.squareup.javapoet.TypeName;
import lombok.Getter;

import java.util.LinkedHashSet;
import java.util.Objects;

@Getter
public final class ElementParameter implements Element {

  private final int index;
  private final String name;
  private final TypeName type;
  private final LinkedHashSet<ElementAnnotation> annotations;

  public ElementParameter(int index, String name, TypeName type, LinkedHashSet<ElementAnnotation> annotations) {
    this.index = index;
    this.name = name;
    this.type = type;
    this.annotations = annotations;
  }

  @Override
  public String keyword() {
    return Utils.getSimpleName(type);
  }

  @Override
  public String toString() {
    return keyword();
  }

  @Override
  public boolean equals(Object o) {
    if (o == null || getClass() != o.getClass()) return false;
    ElementParameter parameter = (ElementParameter) o;
    return index == parameter.index && Objects.equals(type, parameter.type);
  }

  @Override
  public int hashCode() {
    return Objects.hash(index, type);
  }
}

File: src/main/kotlin/cn/labzen/web/ap/internal/element/ElementParameter.kt
package cn.labzen.web.ap.internal.element

import cn.labzen.web.ap.internal.Utils
import com.squareup.javapoet.TypeName

data class ElementParameter(
  val index: Int,
  val name: String,
  val type: TypeName,
  val annotations: LinkedHashSet<ElementAnnotation>,
) : Element {

  override fun keyword(): String =
    Utils.getSimpleName(type)

  override fun toString(): String =
    keyword()

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as ElementParameter

    if (index != other.index) return false
    if (type != other.type) return false

    return true
  }

  override fun hashCode(): Int {
    var result = index
    result = 31 * result + type.hashCode()
    return result
  }
}